src/firewall.ts

```diff
@@ -460,147 +469,300 @@ class WirefilterFirewallRule implements FirewallRule {
     }
 
     match(req: Request): boolean {
-        const wirefilter = this.wirefilter;
-        const exec_ctx = wirefilter.wirefilter_create_execution_context(this.scheme);
+        const exec_ctx = ExecutionContext.buildFromRequest(this.wirefilter, this.scheme, this.lists, req);
+
+        try {
+            return this.matchUsingContext(exec_ctx);
+        } finally {
+            exec_ctx.free();
+        }
+    }
+
+    matchUsingContext(execCtx: ExecutionContext): boolean {
+        const matchResult = this.wirefilter.wirefilter_match(this.filter, execCtx.internalPointer);
+        if (matchResult.ok.success != 1) {
+            throw Error(`Filter can't be matched: ${matchResult.err.msg.data}`);
+        }
+        return matchResult.ok.value;
+    }
+}
+
+
+type Phase = 'http_ratelimit' | 'http_request_sbfm' | 'http_request_firewall_managed'
+type Product = 'zoneLockdown' | 'uaBlock' | 'bic' | 'hot' | 'securityLevel' | 'rateLimit' | 'waf'
+
+type FirewallAction =
+    | {
+        type: 'skip'
+        ruleset?: 'current'
+        phases?: Phase[]
+        products?: Product[]
+    }
+    | {
+        type: 'block' | 'challenge' | 'managed_challenge' | 'js_challenge' | 'log'
+    }
+
+type Rule = {
+    id: string
+    expression: WirefilterFirewallRule
+    action: FirewallAction
+}
+
+type RulesetMatchResult = {
+    loggedRules: string[]
+    skippedPhases: Set<Phase>
+    skippedProducts: Set<Product>
```

Comment:
Think rather than this should just return id + action for each matched rule so that we can get the exact list of matched rules.

If we want to get these could wrap it in some other object type and add some helpers but feel its more useful to expose the exact matches that were made as part of the execution.

Reply:
To me, that couples the tests much more tighter to the implementation of the rules, which makes writing these tests more brittle, reducing their usefulness. 

Adding a list of all matched rules and their actions as well to this gives it flexibility to test based on what rules were triggered specifically, but to me this is targeting integration tests, so we should be side-stepping the details and focusing on the behaviours.

Reply:
I don't believe we should be tailoring this open source runtime against how we personally want to run our tests specifically and should make an effort to match the behaviour/outputs of the actual rulesets implementation as much as possible.

Would be very easy to add helper wrappers either internally or in this repo to achieve the above but trying to do the reverse and get the full list of matched rules would be impossible outside of forking this repo.

Reply:
Could you give me some examples of what you'd expect it to behave like? e.g does the ruleset matching terminate when it hits a block rule? or just a simple list of what matched? Something like:
```ts
match(req: Request): Array<{ruleId: string, action: Action}> {
  const execCtx = ExecutionContext.buildFromRequest(request);
  const matches = []
  for(const rule of this.rules) {
     if(rule.expression.matchUsingContext(execCtx)) {
      matches.push({ ruleId: rule.id, action: rule.action })
    } 
  }
  return matches
}
```

Reply:
I expect it to behave exactly the same as it does in the real system which you can infer from [logs](https://developers.cloudflare.com/logs/reference/log-fields/zone/http_requests/) and the docs around the various [actions](https://developers.cloudflare.com/ruleset-engine/rules-language/actions/). 

As an example in the logs the rules appear as ordered list of all the actions matched with a terminating action appearing as the last action in said list. You can also get a terminating action out via the logs but for our use feel we can just add that as a helper rather than including it as its own separate data field in the result.

This structure should be adopted at the lowest level her but nothing stopping that being returned as part of a class that also add helpers to get stuff like what external stages were skipped, last action, terminated early and other stuff.

Reply:
Alright, have updated and added the matches with helper functions to extract the higher level details.